### PR TITLE
dc_threaded_queue

### PR DESCRIPTION
**Pull request by KarelChanivecky**
Hi team, please fo over this and let me know if you have any suggestions. I for once wasn't sure if I wanted to name it dc_threaded_queue given that I don't like exposing implementation in the identifiers. However I wanted to leave dc_threaded in case a different queue implementation is created at some time.

### CODE REVIEW

File: lib/dc_threaded_queue.c [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+void close_sem( size_t sem_id ) {
+    char name[255] = { 0 };
+    get_sem_name( name, sem_id );
+    sem_unlink( name );
+}
+
+
+int add( struct dc_threaded_queue * queue, void * element );
+
+void * take( struct dc_threaded_queue * queue );
```
Comment (agoryelov) [inline lib/dc_threaded_queue.c:59]: Should these go into the header?

File: lib/dc_threaded_queue.h [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+* The dc_threaded_queue provides a thread safe queue.
+*
+* Synchronization is implemented using the POSIX standard.
+* Hence, this library is not compatible with Windows as of Apr 9, 2021.
+*
+* This queue will provide constant time for adding and taking.
+*
+* To use:
+* 1- initialize using: dc_threaded_queue_init
+* 2- add elements by: <queue_identifier>.add(<element);
```
Comment (agoryelov) [inline lib/dc_threaded_queue.h:14]: <element>

File: lib/dc_threaded_queue.h [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+#include "pthread.h"
+#include "dlinked_list.h"
+#include "semaphore.h"
+
+// sem consumer_wait
+#define DCTQ_ERR_SWAIT -10
+
+// sem post
+#define DCTQ_ERR_SPOST -11 //
```
Comment (agoryelov) [inline lib/dc_threaded_queue.h]: //

**Review by agoryelov — COMMENTED** [review]
Looks good, maybe a bit of clean up needed. 

Also would be nice to have usage docs for all the data structures together in an .md file somewhere.

**Comment by KarelChanivecky** [comment]
fixed all the issues outlined by @agoryelov

File: lib/dc_threaded_queue.h (hunk 2 above)
Comment (KarelChanivecky) [reply lib/dc_threaded_queue.h:14]: what's the issue here?

File: lib/dc_threaded_queue.c (hunk 1 above)
Comment (KarelChanivecky) [reply lib/dc_threaded_queue.c:59]: I am encapsulating the behaviour as part of the struct in this datastructure. Experimenting on how to control namespace pollution.

**Review by agoryelov — APPROVED** [review]
LGTM

**Pull request merged**